### ISSUE TITLE
Different results with MPI

### ISSUE THREAD

**Issue by keiyamamo** [KVSlab/Aneurysm_Workflow_FSI#60]
https://github.com/KVSlab/Aneurysm_Workflow_FSI/blob/f43e1a3a34db69c79454e452ead2846e0b0fe84d/src/fsipy/simulations/simulation_common.py#L191-L211

While I was working on #59, I found that the function evaluation with Probe points are different when running with different number of cores. I will see if I can fix the issue.

**Comment by keiyamamo** [comment]
There are two reasons why this does not work on parallel. For the `peval` function to work, there are two underlying assumptions. 

1. Point `x` is at the vertex and doesn’t require extrapolation
2. Function does not allow extrapolation `f.get_allow_extrapolation() = False`

With these two conditions met, the processor that has the vertex in it can evaluate the function without extrapolation and the other processors return `yloc = inf`. Then MPI `Allreduce` picks the minimum value (which is always the value from the processor that has probe point in it), and we get the correct evaluation. 

In our case, `probe_points` are generally not on the vertex, and thus we need to use extrapolation to get the value. On the other hand, if we allow extrapolation, all the processors return some value ( $\neq$ inf) and `Allreduce` does not work as expected.

**Comment by jorgensd** [comment]
> Point `x` is at the vertex and doesn’t require extrapolation

I don't agree with this. I think the assumption is that the point is within a cell (not necessarily at a vertex). Then it won't need extrapolation for the cell in question.

**Comment by keiyamamo** [comment]
Hi @jorgensd 
I’m a bit confused and would like to ask your help. 
Essentially, we just want to evaluate a function at some point in the domain on parallel and I can see that the function works as expected with simple cases (UnitSquare and so on). However, somehow with the mesh I’m using (3D stenosis mesh), it’s either that all the processors can evaluate the function or none of them can evaluate the function. 

Here is the output of evaluated value from 2 processors. 
with `f.set_allow_extrapolation(True)`
```
yloc = -77.11148506839527 on rank 0
yloc = -2.256944284811986e-09 on rank 1
```
with `f.set_allow_extrapolation(False)`
```
yloc = inf on rank 0
yloc = inf on rank 1
```

Does this mean that the points I want to evaluate at exist in both processors?

**Comment by keiyamamo** [comment]
> > Point `x` is at the vertex and doesn’t require extrapolation
> 
> I don't agree with this. I think the assumption is that the point is within a cell (not necessarily at a vertex). Then it won't need extrapolation for the cell in question.
> 
> However if the point is on a boundary (say an exterior facet), it might be outside the cell by machine precision, and then all cells needs to extrapolate.

You’re right. I figured out that the point does not need to be at vertex, but still not sure why it doesn’t work (see above)

**Comment by keiyamamo** [comment]
Adding a bit more information. The output of 
```
 collied_index = mesh.bounding_box_tree().compute_first_collision(Point(x))
 print(f"collied_index = {collied_index} on rank {MPI.rank(MPI.comm_world)}")
``
is 
```
collied_index = 4294967295 on rank 0
collied_index = 4294967295 on rank 1
```
So I’m guessing both processors don’t recognize the points?

**Comment by jorgensd** [comment]
We can have a look on monday.

**Comment by keiyamamo** [comment]
We figured out that the probe points were actually not inside the domain. This is due to the fact that we make probe points based on the centerline, which is not scaled, while the h5 mesh was saved with scaling factor 0.001. 

To avoid this issue, we should either work with scaled surface https://github.com/KVSlab/Aneurysm_Workflow_FSI/blob/bd54489fb1d7581f6b79631ca93bdcdb5dc85bbd/src/fsipy/automatedPreprocessing/automated_preprocessing.py#L118-L119 

or 

scale probe points if the mesh was saved with scaling.

**Comment by keiyamamo** [comment]
Fixed with #77
